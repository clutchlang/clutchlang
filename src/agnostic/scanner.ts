import { splitLines } from './strings';

/**
 * Represents a section of @member text.
 *
 * Contains enough information in order to represent actionable error or
 * informational messages. Some members in some implementations may be lazy.
 */
export class SourceSpan {
  private static readonly matchNewLine = /\n|\r/;

  constructor(
    public readonly offset: number,
    public readonly column: number,
    public readonly line: number,
    public readonly text: string
  ) {
    if (offset < 0) {
      throw new RangeError(`Invalid offset: ${offset}`);
    }
    if (column < 0) {
      throw new RangeError(`Invalid column: ${column}`);
    }
    if (line < 0) {
      throw new RangeError(`Invalid line: ${line}`);
    }
  }

  /**
   * Length of the span of information.
   */
  public get length(): number {
    return this.text.length;
  }

  /**
   * All lines, separated by line number and text, present in @member text.
   *
   * **NOTE**: `\n\r` is interpreted as a single line break, not two.
   */
  public get lines(): Array<{ line: number; text: string }> {
    return this.isMultiLine
      ? this.computeMultiLines()
      : [
          {
            line: this.line,
            text: this.text,
          },
        ];
  }

  /**
   * Whether the @member text spans over multiple lines.
   *
   * When true, @member lines will represent multiple line numbers.
   */
  public get isMultiLine(): boolean {
    return SourceSpan.matchNewLine.test(this.text);
  }

  private computeMultiLines(): Array<{ line: number; text: string }> {
    return Array.from(
      splitLines(this.text).map((text, index) => {
        return {
          line: this.line + index,
          text,
        };
      })
    );
  }
}

/**
 * A function that handles receiving a @param span and @param error.
 *
 * May optionally return token(s) to recover from the error. Otherwise it is
 * expected the error reporting was fatal and that parsing should be immediately
 * stopped.
 */
export type ErrorReporter = (
  span: SourceSpan,
  error: ScanningError
) => Token | Token[] | void;

/**
 * Error types that may be provided to a @see ErrorReporter.
 */
export type ScanningError = UnexpectedTokenError;

/**
 * A syntax error that occurs as a result of an unexpected @member token.
 *
 * Optionally contains a @member context, or what was being parsed when the
 * error ocurrred, i.e. "argument list" or "function body". May be omitted if
 * the context is not clear or provided.
 */
export class UnexpectedTokenError extends SyntaxError {
  constructor(public readonly token: Token, public readonly context?: string) {
    super(`Unexpected "token"${context ? ` in ${context}` : ''}.`);
  }
}

/**
 * A token that was scanned.
 */
export class Token {}
