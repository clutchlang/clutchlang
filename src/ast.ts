export * from './language/ast/token';
