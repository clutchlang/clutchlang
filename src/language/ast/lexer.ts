export * from './lexer/token';
export * from './lexer/tokenizer';
