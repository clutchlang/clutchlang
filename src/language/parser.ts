export * from './parser/expressions';
export * from './parser/factory';
export * from './parser/nodes';
export * from './parser/operators';
export * from './parser/statements';
export * from './parser/visitors';
