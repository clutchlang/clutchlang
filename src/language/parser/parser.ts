import { FileParser } from './parse/file';

export class ClutchParser extends FileParser {}
