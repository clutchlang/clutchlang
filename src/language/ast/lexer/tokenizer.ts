import { StringScanner } from '../../../agnostic/scanner';
import {
  Characters,
  isDigit,
  isHexadecimal,
  isLetter,
  isWhiteSpace,
} from '../../../agnostic/strings';
import * as tokens from './token';

function isIdentifier(character: number): boolean {
  return isIdentifierStart(character) || isDigit(character);
}

function isIdentifierStart(character: number): boolean {
  return isLetter(character) || character === Characters.$_;
}

/**
 * Tokenizes and returns @param program as a series of @see Token.
 */
export function tokenize(
  program: string,
  onError?: InvalidTokenReporter
): tokens.Token[] {
  return new ClutchLexer().scanProgram(new StringScanner(program), onError);
}

/**
 * Reports a @param message at the provided @param span.
 */
export type InvalidTokenReporter = (offset: number, length: number) => void;

/**
 * A language-level lexer that yields tokens understood by the grammar.
 */
export class ClutchLexer {
  public static readonly keywords: {
    [index: string]: tokens.IKeywordTokenType;
  } = {
    const: tokens.$Const,
    else: tokens.$Else,
    external: tokens.$External,
    false: tokens.$False,
    if: tokens.$If,
    let: tokens.$Let,
    return: tokens.$Return,
    then: tokens.$Then,
    true: tokens.$True,
    type: tokens.$Type,
  };

  private program!: StringScanner;
  private onError!: InvalidTokenReporter;
  private position!: number;
  private lastComments: tokens.Token[] = [];

  public scanProgram(
    program: StringScanner,
    onError: InvalidTokenReporter = (offset, length) => {
      const span = this.program.source.span(offset, offset + length);
      throw new SyntaxError(
        `Invalid token "${span.text}" at ${span.line}:${span.column}`
      );
    }
  ): tokens.Token[] {
    this.position = 0;
    this.program = program;
    this.onError = onError;
    const scanned: tokens.Token[] = [];
    while (program.hasNext()) {
      const token = this.scanToken();
      if (token) {
        scanned.push(token);
      }
    }
    scanned.push(this.createToken(tokens.$EOF, ''));
    return scanned;
  }

  private scanToken(): tokens.Token | undefined {
    const character = this.program.advance();
    switch (character) {
      case Characters.$LPAREN: // "("
        return this.createToken(tokens.$LeftParen);
      case Characters.$RPAREN: // ")"
        return this.createToken(tokens.$RightParen);
      case Characters.$LCURLY: // "{"
        return this.createToken(tokens.$LeftCurly);
      case Characters.$RCURLY: // "}"
        return this.createToken(tokens.$RightCurly);
      case Characters.$COLON: // ":"
        return this.createToken(tokens.$Colon);
      case Characters.$PERIOD: // "."
        return this.createToken(tokens.$Period);
      case Characters.$PLUS: // "+" or "+=" or "++"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? tokens.$PlusEquals
            : this.program.match(Characters.$PLUS)
              ? tokens.$PlusPlus
              : tokens.$Plus
        );
      case Characters.$MINUS: // "-" or "--" or "-=" or "->"
        return this.createToken(
          this.program.match(Characters.$RANGLE)
            ? tokens.$DashRightAngle
            : this.program.match(Characters.$EQUALS)
              ? tokens.$DashEquals
              : this.program.match(Characters.$MINUS)
                ? tokens.$DashDash
                : tokens.$Dash
          // Code coverage considers the `);` to be a missed line :(
          /* istanbul ignore next */
        );
      case Characters.$STAR: // "*" or "*="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? tokens.$StarEquals
            : tokens.$Star
        );
      case Characters.$PERCENT: // "*" or "*="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? tokens.$PercentEquals
            : tokens.$Percent
        );
      case Characters.$EQUALS: // "=" or "==" or "==="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? this.program.match(Characters.$EQUALS)
              ? tokens.$EqualsEqualsEquals
              : tokens.$EqualsEquals
            : tokens.$Equals
        );
      case Characters.$LANGLE: // "<" or "<=" or "<<"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? tokens.$LeftAngleEquals
            : this.program.match(Characters.$LANGLE)
              ? tokens.$LeftAngleLeftAngle
              : tokens.$LeftAngle
        );
      case Characters.$RANGLE: // ">" or ">=" or ">>"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? tokens.$RightAngleEquals
            : this.program.match(Characters.$RANGLE)
              ? tokens.$RightAngleRightAngle
              : tokens.$RightAngle
        );
      case Characters.$SLASH: // "/" or "//" or "/="
        return this.scanSlash();
      case Characters.$SQUOTE: // "'"
        return this.scanString();
      case Characters.$PIPE: // "|" or "||"
        return this.createToken(
          this.program.match(Characters.$PIPE) ? tokens.$PipePipe : tokens.$Pipe
        );
      case Characters.$AND: // "&" or "&&"
        return this.createToken(
          this.program.match(Characters.$AND) ? tokens.$AndAnd : tokens.$And
        );
      case Characters.$TILDE: // "~"
        return this.createToken(tokens.$Tilde);
      case Characters.$CARET: // "^"
        return this.createToken(tokens.$Caret);
      case Characters.$EXCLAIM: // "!" or "!=" or "!==""
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? this.program.match(Characters.$EQUALS)
              ? tokens.$ExclaimEqualsEquals
              : tokens.$ExclaimEquals
            : tokens.$Exclaim
        );
      default:
        if (isWhiteSpace(character)) {
          this.position = this.program.position;
          break;
        }
        if (isDigit(character)) {
          return this.scanNumber(character);
        }
        if (isIdentifierStart(character)) {
          return this.scanIdentifierOrKeyword();
        }
        this.reportError();
    }
    return;
  }

  /**
   * Scans a string, returning `undefined` if there was no string.
   */
  private scanString(): tokens.Token {
    while (this.program.hasNext()) {
      if (this.program.match(Characters.$SQUOTE)) {
        this.position++;
        let lexeme = this.substring();
        lexeme = lexeme.substring(0, lexeme.length - 1);
        return this.createToken(tokens.$String, lexeme);
      }
      this.program.advance();
    }
    this.reportError();
    this.position++;
    return this.createToken(tokens.$String);
  }

  /**
   * Scans any tokens that are valid numeric literals
   */
  private scanNumber(starting: number): tokens.Token {
    if (starting === Characters.$0) {
      if (
        this.program.match(Characters.$X) ||
        this.program.match(Characters.$x)
      ) {
        return this.scanHexNumber();
      }
    }
    if (
      this.program.match(Characters.$E) ||
      this.program.match(Characters.$e)
    ) {
      return this.scanDigits();
    }
    while (this.program.hasNext()) {
      const next = this.program.peek();
      if (this.program.match(Characters.$PERIOD)) {
        break;
      }
      if (!isDigit(next)) {
        break;
      }
      this.program.advance();
    }
    return this.scanDigits();
  }

  /**
   * Scans any tokens that are valid hexadecimel digits, returning a number.
   */
  private scanHexNumber(): tokens.Token {
    this.scanPredicate(isHexadecimal);
    return this.createToken(tokens.$Number);
  }

  /**
   * Scans any tokens that are valid digits, returning a number.
   */
  private scanDigits(): tokens.Token {
    this.scanPredicate(isDigit);
    return this.createToken(tokens.$Number);
  }

  /**
   * Scans any tokens that are valid identifiers.
   */
  private scanIdentifierOrKeyword(): tokens.Token {
    this.scanPredicate(isIdentifier);
    const identifierOrKeyword = this.substring();
    return new tokens.Token(
      this.position - identifierOrKeyword.length,
      ClutchLexer.keywords[identifierOrKeyword] || tokens.$Identifier,
      this.clearComments(),
      identifierOrKeyword
    );
  }

  /**
   * Scans tokens preceding with "/".
   */
  private scanSlash(): tokens.Token | undefined {
    // "//"
    if (this.program.match(Characters.$SLASH)) {
      this.advanceToEOL();
      this.lastComments.push(
        new tokens.Token(
          this.position,
          tokens.$Comment,
          [],
          this.substring().trim()
        )
      );
      return;
    }
    // "/" or "/="
    return this.createToken(
      this.program.match(Characters.$EQUALS)
        ? tokens.$SlashEquals
        : tokens.$Slash
    );
  }

  /**
   * Scans until the EOF or until @param predicate returns false.
   */
  private scanPredicate(predicate: (character: number) => boolean): void {
    while (this.program.hasNext()) {
      const peek = this.program.peek();
      if (!predicate(peek)) {
        break;
      }
      this.program.advance();
    }
  }

  /**
   * Advances the position of the scanner until either EOF or EOL.
   *
   * **NOTE**: Treats `CR+LF` as a DOS-stlye line ending, not two lines.
   */
  private advanceToEOL(): void {
    while (this.program.hasNext()) {
      // "\n" (LF)
      if (this.program.match(Characters.$LF)) {
        return;
      }
      // "\r" (CR)
      if (this.program.match(Characters.$CR)) {
        // "\r\n" (CR+LF)
        this.program.match(Characters.$LF);
        return;
      }
      this.program.advance();
    }
  }

  /**
   * Returns a new token with a @param kind and the current substring.
   */
  private createToken(
    kind: tokens.ITokenTypes,
    content = this.substring()
  ): tokens.Token {
    return new tokens.Token(
      this.position - content.length,
      kind,
      this.clearComments(),
      content
    );
  }

  /**
   * Clear and retrieve comments lexed.
   */
  private clearComments(): tokens.Token[] {
    const comments = this.lastComments;
    this.lastComments = [];
    return comments;
  }

  /**
   * Returns the current substring, advancing the position counter.
   */
  private substring(): string {
    const start = this.position;
    const end = this.program.position;
    this.position = end;
    return this.program.substring(start, end);
  }

  /**
   * Reports a lexical @param error occurring at @param offset.
   */
  private reportError(offset = this.position, length = 1): void {
    this.onError(offset, length);
  }
}
