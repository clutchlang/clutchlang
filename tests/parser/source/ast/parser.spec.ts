// tslint:disable:no-magic-numbers

import {
  AstCompilationUnit,
  AstInvocationExpression,
  AstLiteralIdentifier,
} from '../../../../src/parser/source/ast/node';
import { AstParser } from '../../../../src/parser/source/ast/parser';
import { TokenScanner } from '../../../../src/parser/source/ast/scanner';
import { Lexer } from '../../../../src/parser/source/tokenizer/lexer';
import { SourceScanner } from '../../../../src/parser/source/tokenizer/scanner';

function parse(program: string): AstCompilationUnit {
  const scanner = new SourceScanner(program);
  const lexer = new Lexer(scanner);
  const tokens = Array.from(lexer);
  const parser = new AstParser(new TokenScanner(tokens));
  return parser.parseCompilationUnit();
}

describe('AstParser', () => {
  it('should parse a blank program with two functions', () => {
    const unit = parse(`
      one => {}
      two => {}
    `);
    expect(unit.functions).toHaveLength(2);
    expect(unit.functions[0].name).toBe('one');
    expect(unit.functions[0].body).toHaveLength(0);
    expect(unit.functions[1].name).toBe('two');
    expect(unit.functions[1].body).toHaveLength(0);
  });

  it('should parse a lambda-style function', () => {
    const unit = parse(`
      main => true
    `);
    expect(unit.functions).toHaveLength(1);
  });

  it('should parse an expression-body function', () => {
    const unit = parse(`
      main => {
        1
        1.5
        -1
        -1.5
        'Hello'
        true
        false
        foobar
      }
    `);
    expect(unit.functions).toHaveLength(1);
    // Poor man's stringify.
    // TODO: Remove and replace with a visitor/humanizer.
    expect(
      unit.functions[0].body.map(e => {
        if (e instanceof AstLiteralIdentifier) {
          return e.name;
        }
        if (e instanceof AstInvocationExpression) {
          return `${e.target}(...)`;
        }
        return e.value;
      })
    ).toEqual([1, 1.5, -1, -1.5, 'Hello', true, false, 'foobar']);
  });

  it('should parse a function with an invocation', () => {
    const unit = parse(`
      main => {
        print('Hello World')
      }
    `);
    expect(unit.functions).toHaveLength(1);
  });
});
