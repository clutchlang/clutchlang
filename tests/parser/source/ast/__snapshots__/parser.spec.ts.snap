// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AstNode.beginToken/endToken should be set 1`] = `
Array [
  Object {
    "kind": "Identifier",
    "offset": "5 -> 9",
    "value": "main",
  },
  Object {
    "kind": "}",
    "offset": "379 -> 380",
    "value": "}",
  },
  Object {
    "kind": "Identifier",
    "offset": "5 -> 9",
    "value": "main",
  },
  Object {
    "kind": "}",
    "offset": "59 -> 60",
    "value": "}",
  },
  Object {
    "kind": "Identifier",
    "offset": "21 -> 24",
    "value": "rfn",
  },
  Object {
    "kind": ")",
    "offset": "25 -> 26",
    "value": ")",
  },
  Object {
    "kind": "Identifier",
    "offset": "33 -> 36",
    "value": "fn1",
  },
  Object {
    "kind": ")",
    "offset": "38 -> 39",
    "value": ")",
  },
  Object {
    "kind": "Identifier",
    "offset": "37 -> 38",
    "value": "a",
  },
  Object {
    "kind": "Identifier",
    "offset": "37 -> 38",
    "value": "a",
  },
  Object {
    "kind": "Identifier",
    "offset": "46 -> 49",
    "value": "fn2",
  },
  Object {
    "kind": ")",
    "offset": "53 -> 54",
    "value": ")",
  },
  Object {
    "kind": "Identifier",
    "offset": "50 -> 51",
    "value": "a",
  },
  Object {
    "kind": "Identifier",
    "offset": "50 -> 51",
    "value": "a",
  },
  Object {
    "kind": "Identifier",
    "offset": "52 -> 53",
    "value": "b",
  },
  Object {
    "kind": "Identifier",
    "offset": "52 -> 53",
    "value": "b",
  },
  Object {
    "kind": "Identifier",
    "offset": "66 -> 69",
    "value": "rfn",
  },
  Object {
    "kind": "Boolean",
    "offset": "73 -> 77",
    "value": "true",
  },
  Object {
    "kind": "Boolean",
    "offset": "73 -> 77",
    "value": "true",
  },
  Object {
    "kind": "Boolean",
    "offset": "73 -> 77",
    "value": "true",
  },
  Object {
    "kind": "Identifier",
    "offset": "83 -> 86",
    "value": "efn",
  },
  Object {
    "kind": "}",
    "offset": "114 -> 115",
    "value": "}",
  },
  Object {
    "kind": "return",
    "offset": "98 -> 104",
    "value": "return",
  },
  Object {
    "kind": "Boolean",
    "offset": "105 -> 109",
    "value": "true",
  },
  Object {
    "kind": "Boolean",
    "offset": "105 -> 109",
    "value": "true",
  },
  Object {
    "kind": "Boolean",
    "offset": "105 -> 109",
    "value": "true",
  },
  Object {
    "kind": "Identifier",
    "offset": "121 -> 124",
    "value": "fn1",
  },
  Object {
    "kind": "}",
    "offset": "338 -> 339",
    "value": "}",
  },
  Object {
    "kind": "Boolean",
    "offset": "139 -> 143",
    "value": "true",
  },
  Object {
    "kind": "Boolean",
    "offset": "139 -> 143",
    "value": "true",
  },
  Object {
    "kind": "Boolean",
    "offset": "150 -> 155",
    "value": "false",
  },
  Object {
    "kind": "Boolean",
    "offset": "150 -> 155",
    "value": "false",
  },
  Object {
    "kind": "Number",
    "offset": "162 -> 166",
    "value": "-1.5",
  },
  Object {
    "kind": "Number",
    "offset": "162 -> 166",
    "value": "-1.5",
  },
  Object {
    "kind": "Number",
    "offset": "173 -> 175",
    "value": "-1",
  },
  Object {
    "kind": "Number",
    "offset": "173 -> 175",
    "value": "-1",
  },
  Object {
    "kind": "Number",
    "offset": "182 -> 183",
    "value": "0",
  },
  Object {
    "kind": "Number",
    "offset": "182 -> 183",
    "value": "0",
  },
  Object {
    "kind": "Number",
    "offset": "190 -> 191",
    "value": "1",
  },
  Object {
    "kind": "Number",
    "offset": "190 -> 191",
    "value": "1",
  },
  Object {
    "kind": "Number",
    "offset": "198 -> 201",
    "value": "1.5",
  },
  Object {
    "kind": "Number",
    "offset": "198 -> 201",
    "value": "1.5",
  },
  Object {
    "kind": "Number",
    "offset": "208 -> 216",
    "value": "0xAABBCC",
  },
  Object {
    "kind": "Number",
    "offset": "208 -> 216",
    "value": "0xAABBCC",
  },
  Object {
    "kind": "String",
    "offset": "223 -> 236",
    "value": "'Hello World'",
  },
  Object {
    "kind": "String",
    "offset": "223 -> 236",
    "value": "'Hello World'",
  },
  Object {
    "kind": "String",
    "offset": "243 -> 256",
    "value": "\\"Hello World\\"",
  },
  Object {
    "kind": "String",
    "offset": "243 -> 256",
    "value": "\\"Hello World\\"",
  },
  Object {
    "kind": "Identifier",
    "offset": "263 -> 269",
    "value": "fooBar",
  },
  Object {
    "kind": ")",
    "offset": "287 -> 288",
    "value": ")",
  },
  Object {
    "kind": "Boolean",
    "offset": "277 -> 282",
    "value": "false",
  },
  Object {
    "kind": "Boolean",
    "offset": "277 -> 282",
    "value": "false",
  },
  Object {
    "kind": "Boolean",
    "offset": "283 -> 287",
    "value": "true",
  },
  Object {
    "kind": "Boolean",
    "offset": "283 -> 287",
    "value": "true",
  },
  Object {
    "kind": "(",
    "offset": "295 -> 296",
    "value": "(",
  },
  Object {
    "kind": ")",
    "offset": "306 -> 307",
    "value": ")",
  },
  Object {
    "kind": "Boolean",
    "offset": "296 -> 300",
    "value": "true",
  },
  Object {
    "kind": "Boolean",
    "offset": "296 -> 300",
    "value": "true",
  },
  Object {
    "kind": "Boolean",
    "offset": "301 -> 306",
    "value": "false",
  },
  Object {
    "kind": "Boolean",
    "offset": "301 -> 306",
    "value": "false",
  },
  Object {
    "kind": "let",
    "offset": "314 -> 317",
    "value": "let",
  },
  Object {
    "kind": "Boolean",
    "offset": "329 -> 333",
    "value": "true",
  },
  Object {
    "kind": "Boolean",
    "offset": "329 -> 333",
    "value": "true",
  },
  Object {
    "kind": "Boolean",
    "offset": "329 -> 333",
    "value": "true",
  },
  Object {
    "kind": "Identifier",
    "offset": "345 -> 348",
    "value": "fn2",
  },
  Object {
    "kind": "}",
    "offset": "379 -> 380",
    "value": "}",
  },
  Object {
    "kind": "Identifier",
    "offset": "365 -> 366",
    "value": "a",
  },
  Object {
    "kind": "Identifier",
    "offset": "365 -> 366",
    "value": "a",
  },
  Object {
    "kind": "Identifier",
    "offset": "373 -> 374",
    "value": "b",
  },
  Object {
    "kind": "Identifier",
    "offset": "373 -> 374",
    "value": "b",
  },
]
`;
